sdk/src/main/java/com/mapbox/maps/renderer/FpsManager.kt

```diff
@@ -0,0 +1,213 @@
+  private var previousFrameTimeNs = -1L
+  private var previousDrawnFrameIndex = 0
+  private var frameRenderTimeAccumulatedNs = 0L
+
+  private var preRenderTimeNs = -1L
+  private var choreographerTicks = 0
+  private var choreographerSkips = 0
+
+  internal var fpsChangedListener: OnFpsChangedListener? = null
+
+  fun setScreenRefreshRate(screenRefreshRate: Int) {
+    if (this.screenRefreshRate == screenRefreshRate) {
+      return
+    }
+    this.screenRefreshRate = screenRefreshRate
+    screenRefreshPeriodNs = ONE_SECOND_NS / screenRefreshRate
+    if (userRefreshRate != USER_DEFINED_REFRESH_RATE_NOT_SET) {
+      userToScreenRefreshRateRatio = userRefreshRate.toDouble() / screenRefreshRate
+      logI(TAG, "User defined ratio is $userToScreenRefreshRateRatio")
+    }
+    if (LOG_STATISTICS) {
+      logI(
+        TAG,
+        "Device screen frequency is $screenRefreshRate," +
+          " user defined refresh rate is ${if (userRefreshRate == USER_DEFINED_REFRESH_RATE_NOT_SET) "not set" else userRefreshRate}"
+      )
+    }
+  }
+
+  fun setUserRefreshRate(refreshRate: Int) {
+    if (userRefreshRate != refreshRate) {
+      userRefreshRate = refreshRate
+      logI(TAG, "User set max FPS to $userRefreshRate")
+      if (screenRefreshRate != SCREEN_METRICS_NOT_DEFINED) {
+        userToScreenRefreshRateRatio = userRefreshRate.toDouble() / screenRefreshRate
+        logI(TAG, "User defined ratio is $userToScreenRefreshRateRatio")
+      }
+    }
+  }
+
+  /**
+   * Return true if rendering should happen this frame and false otherwise.
+   *
+   * @param frameTimeNs time value from Choreographer in [System.nanoTime] timebase.
+   */
+  fun preRender(frameTimeNs: Long): Boolean {
+    // no need to perform neither pacing nor FPS calculation when setMaxFps / setOnFpsChangedListener was not called by the user
+    if (userToScreenRefreshRateRatio == null && fpsChangedListener == null) {
+      return true
+    }
+    // clear any scheduled task as new render call is about to happen
+    handler.removeCallbacksAndMessages(fpsManagerToken)
+    // update frame stats, count skipped / total VSYNC events
+    updateFrameStats(frameTimeNs)
+    userToScreenRefreshRateRatio?.let {
+      return performPacing(it)
+    }
+    return true
+  }
+
+  fun postRender() {
+    val frameRenderTimeNs = System.nanoTime() - preRenderTimeNs
+    frameRenderTimeAccumulatedNs += frameRenderTimeNs
+    // normally we update FPS counter and reset counters once a second
+    if (choreographerTicks >= screenRefreshRate) {
+      calculateFpsAndReset()
+    } else {
+      // however to produce correct values we also update FPS after IDLE_TIMEOUT_MS
+      // otherwise when updating the map after it was IDLE first update will report
+      // huge delta between new frame and last frame (as we're using dirty rendering)
+      handler.postDelayed(
+        VSYNC_COUNT_TILL_IDLE * (screenRefreshPeriodNs / 10.0.pow(6.0)).toLong(),
```

Comment:
no need to calculate it each time
```suggestion
        VSYNC_COUNT_TILL_IDLE * (screenRefreshPeriodNs / ONE_MILLISECOND_NS).toLong(),
```